lib/usbh/usbh_dev_enum.c

```diff
@@ -0,0 +1,169 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "usbh-private.h"
+#include <libopencm3/usbh/helper/ctrlreq.h>
+
+/**
+ * procedure to enumerate a device
+ *  - prepare the device object
+ *  - set address
+ *  - get device descriptor (get ep0 max size)
+ */
+
+#define CONTROL_TIMEOUT 500
+
+static void enum_failed(usbh_device *dev)
+{
+	LOG_LN("failed to enumerate device");
+	usbh_device_invalidate(dev);
+}
+
+static void enum_success(usbh_device *dev)
+{
+	usbh_host *host = dev->host;
+
+	LOG_LN("successfully enumerated device");
+
+	if (host->connected != NULL) {
+		host->connected(dev);
+	}
+}
+
+/**
+ * Callback when the partial descriptor is readed from the device.
+ * @param dev USB Device
+ * @param status Transfer status
+ * @param urb_id URB ID (ignored)
+ * @warning @a desc has only first 8bytes valid
+ */
+static void got_partial_dev_desc(const usbh_transfer *transfer,
+		usbh_transfer_status status, usbh_urb_id urb_id)
+{
+	(void) urb_id;
+
+	usbh_device *dev = transfer->device;
+	if (status != USBH_SUCCESS) {
+		LOG_LN("failed to read partial device descriptor from device");
+		enum_failed(dev);
+		return;
+	}
+
+	LOG_LN("success in reading partial device descriptor from device");
+
+	struct usb_device_descriptor *desc = transfer->data;
+	LOGF_LN("bLength: %"PRIu8, desc->bLength);
+	LOGF_LN("bDescriptorType: 0x%"PRIx8, desc->bDescriptorType);
+	LOGF_LN("bcdUSB: 0x%"PRIx16, desc->bcdUSB);
+	LOGF_LN("bDeviceClass: 0x%"PRIx8, desc->bDeviceClass);
+	LOGF_LN("bDeviceSubClass: 0x%"PRIx8, desc->bDeviceSubClass);
+	LOGF_LN("bDeviceProtocol: 0x%"PRIx8, desc->bDeviceProtocol);
+	LOGF_LN("bMaxPacketSize0: %"PRIu8, desc->bMaxPacketSize0);
+
+	dev->ep0_max_packet_size = desc->bMaxPacketSize0;
+	enum_success(dev);
+}
+
+/**
+ * Callback when an address is assigned to the device.
+ * @param dev USB Device
+ * @param status Transfer status
+ * @param urb_id URB ID (ignored)
+ */
+static void address_assigned(const usbh_transfer *transfer,
+		usbh_transfer_status status, usbh_urb_id urb_id)
+{
+	(void) urb_id;
+
+	usbh_device *dev = transfer->device;
+	if (status != USBH_SUCCESS) {
+		LOG_LN("failed to set address to device");
+		enum_failed(dev);
+		return;
+	}
+
+	LOG_LN("succeeded in set address to device");
+	dev->address = transfer->setup.wValue;
+
+	LOG_LN("trying to read partial device descriptor from device");
+	usbh_ctrlreq_read_dev_desc(dev, dev->host->buffer, 8, got_partial_dev_desc);
+}
+
+/**
+ * Allocate the next device address from the host
+ * @param host USB Host
+ * @return valid non-zero 7bit address
+ * @return zero on failure
+ */
+static uint8_t alloc_device_address(usbh_host *host)
+{
+	for (unsigned i = 0; i < 127; i++) {
```

Comment:
Declaring the variable inside the for loop here, gets a complaint from the compiler.